### PR TITLE
Subgroups and exactness

### PR DESCRIPTION
**Pull request by jarlg**
(Let me know if it's much cleaner to let you rebase master first, and I'll just PR again. Hope this didn't notify all the 6 participants? If so, sorry!)

The first commit (9f8f308) is a polished version of what's on your `subgroup` branch. It proves that paths of subgroups correspond to isomorphisms respecting the inclusions, and that group embeddings are those maps with trivial kernels.

The second commit (7d67c45) relates the above to exactness by proving that a complex `0 -> A -> B` of groups is purely exact if and only if the map `A -> B` is an embedding.

I've left some comments in the code prefixed by `jarlg:`, and here's a comment I'll leave here:

By removing the (unused) import of `Modalities.Identity` in `Truncations.Core`, I could import `Truncations` into `Modalities.Identity` and prove `isequiv_purely_conn`.

### CODE REVIEW

File: theories/Algebra/Ext/Ext.v [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+  - intro x; cbn.
+    exact (grp_homo_unit f).
+  - apply path_ishprop.
+Defined.
+
+Local Existing Instance ishprop_phomotopy_hset.
+Local Existing Instance isequiv_purely_conn.
+Local Existing Instance ishprop_isexact_hset.
+
+(** The complex 0 -> A -> B is purely exact if and only if f is a group embedding. *)
```
Comment (jarlg) [inline theories/Algebra/Ext/Ext.v:31]: I've improved this comment to "A complex 0 -> A -> B of groups is purely exact if and only if the map A -> B is a group embedding."

File: theories/Algebra/Groups/Group.v [hunk 2]
```diff
@@ -70,6 +71,8 @@ Proof.
   apply left_identity.
 Defined.
 
+(** ** Group algebra *)
```
Comment (jdchristensen) [inline theories/Algebra/Groups/Group.v:74]: A "group algebra" is the algebra formed by taking linear combinations of group elements, so maybe rephrase.  "Moving things around in equations"?

File: theories/Algebra/Groups/Subgroup.v [hunk 3]
```diff
@@ -35,6 +43,102 @@ Coercion subgroup_group : Subgroup >-> Group.
...
+Definition maximal_subgroup {G} : Subgroup G
+  := Build_Subgroup G G _.
+
+(** ** Characterization of paths between subgroups *)
+
+(* jdc: It takes too long for Coq to process the *statement* of this Lemma.
+I think you need to supply some implicit arguments to fix this. 
+I think the lemma might be ok here, since it is used twice below. *)
+(* jarlg: The processing of the statement is instant on my
+computer. What am I missing? Making the lemma local as the name isn't
```
Comment (jdchristensen) [inline theories/Algebra/Groups/Subgroup.v:89]: It's fine for me now.  Not sure what changed.

File: theories/Algebra/Groups/Kernel.v [hunk 4]
```diff
@@ -68,3 +69,70 @@ Proof.
...
+    change mon_unit with (pr1 ((mon_unit; grp_homo_unit f) : grp_kernel f)).
+    pose proof (ap (fun g => g *  -(f a')) (p @ q^)) as P; cbn in P.
+    rewrite right_inverse in P.
+    rewrite <- (grp_homo_inv f) in P.
+    rewrite <- (grp_homo_op f) in P.
+    change (a * -a') with (pr1 ((a * -a'; P) : grp_kernel f)).
+    apply (ap pr1).
+    pose (phi' := ap (@group_type o (subgroup_group A)) phi); cbn in phi'.
+    apply (equiv_ap' (equiv_path _ _ phi')).
+    apply path_ishprop.
```
Comment (jdchristensen) [inline theories/Algebra/Groups/Kernel.v:125]: The contents of the first `-` bullet can be replaced with:

```coq
  - intros phi b.
    apply hprop_inhabited_contr; intros a.
    srapply (contr_equiv' _ (equiv_grp_hfiber _ _ a)^-1%equiv).
    rapply (transport Contr (x:=grp_trivial)).
    exact (ap (@group_type o (subgroup_group A)) phi)^.
```

Not only is it shorter, it avoids `pose`, `rewrite` and `change`, which I think is good.

**Comment by jdchristensen** [comment]
Any idea why the build is failing.  Is it just because I still had an outdated Coq, or something like that?

Feel free to send a PR upstream when you've handled the minor issues I mentioned.

**Comment by jdchristensen** [comment]
All my builds, even of a fresh master, are failing.  It's something to do with a security issue in github, so doesn't mean anything about this PR.

**Comment by Alizter** [comment]
Sorry about this. See https://github.com/HoTT/HoTT/issues/1401, I'll try and have it fixed by today.

**Comment by Alizter** [comment]
Here is the fix https://github.com/HoTT/HoTT/pull/1402

File: theories/Algebra/Groups/Group.v (hunk 2 above)
Comment (jarlg) [reply theories/Algebra/Groups/Group.v:74]: Fair enough! I was thinking of the verb form, analogous to doing "path algebra". But even that isn't a commonly used term in the library. Making it "Working with equations in groups".

File: theories/Algebra/Groups/Kernel.v (hunk 4 above)
Comment (jarlg) [reply theories/Algebra/Groups/Kernel.v:125]: Nice!

File: theories/Algebra/Groups/Group.v (hunk 2 above)
Comment (Alizter) [reply theories/Algebra/Groups/Group.v:74]: Algebra in groups also works

**Comment by jarlg** [comment]
Thanks!
